ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/PushPhysicalPlanBuilder.java

```diff
@@ -0,0 +1,169 @@
+ */
+// CHECKSTYLE_RULES.OFF: ClassDataAbstractionCoupling
+public class PushPhysicalPlanBuilder {
+  // CHECKSTYLE_RULES.ON: ClassDataAbstractionCoupling
+
+  private final ProcessingLogContext processingLogContext;
+  private final PersistentQueryMetadata persistentQueryMetadata;
+  private final Stacker contextStacker;
+  private final QueryId queryId;
+
+  public PushPhysicalPlanBuilder(
+      final ProcessingLogContext processingLogContext,
+      final PersistentQueryMetadata persistentQueryMetadata
+  ) {
+    this.processingLogContext = Objects.requireNonNull(
+        processingLogContext, "processingLogContext");
+    this.persistentQueryMetadata = Objects.requireNonNull(
+        persistentQueryMetadata, "persistentQueryMetadata");
+    this.contextStacker = new Stacker();
+    queryId = uniqueQueryId();
+  }
+
+  /**
+   * Visits the logical plan top-down to build the physical plan.
+   * @param logicalPlanNode the logical plan root node
+   * @return the root node of the tree of physical operators
+   */
+  public PushPhysicalPlan buildPushPhysicalPlan(
+      final LogicalPlanNode logicalPlanNode,
+      final Context context
+  ) {
+    PushDataSourceOperator dataSourceOperator = null;
+
+    final OutputNode outputNode = logicalPlanNode.getNode()
+        .orElseThrow(() -> new IllegalArgumentException("Need an output node to build a plan"));
+
+    if (!(outputNode instanceof KsqlBareOutputNode)) {
+      throw new KsqlException("Push queries expect the root of the logical plan to be a "
+          + "KsqlBareOutputNode.");
+    }
+    // We skip KsqlBareOutputNode in the translation since it only applies the LIMIT
+    PlanNode currentLogicalNode = outputNode.getSource();
+    AbstractPhysicalOperator prevPhysicalOp = null;
+    AbstractPhysicalOperator rootPhysicalOp = null;
+    while (true) {
+      AbstractPhysicalOperator currentPhysicalOp = null;
+      if (currentLogicalNode instanceof PullProjectNode) {
+        currentPhysicalOp = translateProjectNode((PullProjectNode)currentLogicalNode);
+      } else if (currentLogicalNode instanceof PullFilterNode) {
+        currentPhysicalOp = translateFilterNode((PullFilterNode) currentLogicalNode);
+      } else if (currentLogicalNode instanceof DataSourceNode) {
+        currentPhysicalOp = translateDataSourceNode(
+            (DataSourceNode) currentLogicalNode);
+        dataSourceOperator = (PushDataSourceOperator) currentPhysicalOp;
+      } else {
+        throw new KsqlException(String.format(
+            "Error in translating logical to physical plan for scalable push queries:"
+                + " unrecognized logical node %s.", currentLogicalNode));
+      }
+
+      if (prevPhysicalOp == null) {
+        rootPhysicalOp = currentPhysicalOp;
+      } else {
+        prevPhysicalOp.addChild(currentPhysicalOp);
+      }
+      prevPhysicalOp = currentPhysicalOp;
+      // Exit the loop when a leaf node is reached
+      if (currentLogicalNode.getSources().isEmpty()) {
```

Comment:
Just curious, is it true that only `DataSourceNode` falls in this case today?

Reply:
Yeah.  Once you get to the DataSourceNode, that should be it, at least in our current setup.